web/src/test/auto/headless/engine/interfaces/prediction/predictionContext.spec.js

```diff
@@ -118,6 +118,65 @@ describe("PredictionContext", () => {
     assert.equal(suggestions.find((obj) => obj.transform.deleteLeft != 0).displayAs, 'apps');
   });
 
+  it('ignores outdated predictions', async function () {
+    const langProcessor = new LanguageProcessor(worker, new TranscriptionCache());
+    await langProcessor.loadModel(appleDummyModel);  // await:  must fully 'configure', load script into worker.
+
+    const kbdProcessor = new KeyboardProcessor(deviceSpec);
+    const predictiveContext = new PredictionContext(langProcessor, kbdProcessor);
+
+    let updateFake = sinon.fake();
+    predictiveContext.on('update', updateFake);
+
+    let mock = new Mock("appl", 4); // "appl|", with '|' as the caret position.
+    const initialMock = Mock.from(mock);
+    const promise = predictiveContext.setCurrentTarget(mock);
+
+    // Initial predictive state:  no suggestions.  context.initializeState() has not yet been called.
+    assert.equal(updateFake.callCount, 1);
+    assert.isEmpty(updateFake.firstCall.args[0]); // should have no suggestions. (if convenient for testing)
+
+    await promise;
+    let suggestions;
+
+    // Initialization results:  our first set of dummy suggestions.
+    assert.equal(updateFake.callCount, 2);
+    suggestions = updateFake.secondCall.args[0];
+    assert.deepEqual(suggestions.map((obj) => obj.displayAs), ['apple', 'apply', 'apples']);
+    assert.isNotOk(suggestions.find((obj) => obj.tag == 'keep'));
+    assert.isNotOk(suggestions.find((obj) => obj.transform.deleteLeft != 0));
+
+    const baseTranscription = mock.buildTranscriptionFrom(initialMock, null, true);
+
+    // Mocking:  corresponds to the second set of mocked predictions - round 2 of
+    // 'apple', 'apply', 'apples'.
+    const skippedPromise = langProcessor.predict(baseTranscription, kbdProcessor.layerId);
+
+    mock.insertTextBeforeCaret('e'); // appl| + e = apple
+    const finalTranscription = mock.buildTranscriptionFrom(initialMock, null, true);
+
+    // Mocking:  corresponds to the third set of mocked predictions - 'applied'.
+    const expectedPromise = langProcessor.predict(finalTranscription, kbdProcessor.layerId);
+
+    await Promise.all([skippedPromise, expectedPromise]);
+    const expected = await expectedPromise;
+
+    // Despite two predict calls, we should only increase the counter by ONE - we ignore
+    // the 'outdated' / 'skipped' round because it could not respond before its followup.
+    assert.equal(updateFake.callCount, 3);
+    suggestions = updateFake.thirdCall.args[0];
+
+    // This does re-use the apply-revert oriented mocking.
+    // Should skip the (second) "apple", "apply", "apps" round, as it became outdated
+    // by its following request before its response could be received.
+    assert.deepEqual(suggestions.map((obj) => obj.displayAs), ['“apple”', 'applied']);
```

Comment:
Why do we get a different set of suggestions than in the previous test? Ok, the difference is the additional predict call, but why/how would that get triggered in production?

Reply:
> but why/how would that get triggered in production?

- Suppose I quickly double-tap a key on the OSK.
- Or that I am using multiple fingers to type (especially with an Android external keyboard) and type two keys near-simultaneously
- Or am using an older Android device where we previously had rapid-typing bugs due to slow keystroke processing - the predictive-text processing would likely also be slow there.
    - The JS task queue and microtask queue tend to be FIFO in my experience
    - All pending keys to be processed would each trigger their own `predict` call when processed, each one likely cancelling the previous one.
        - Even if the worker does get processing time in parallel, the worker's _response_ would only _then_ fulfill the prediction `Promise`, thus queuing the response, _after_ all the `Promise`s for the pending keys.
        - This, in turn, would also skip the predictions - which is a good thing here.  We don't want to rapidly rotate through numerous invalid suggestions while we're already lagging - updating the DOM does take up CPU time, so we should only do it once the suggestions are stable.

That list isn't necessarily exhaustive, but highlights the most common cases I'd expect to trigger this.